src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,469 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+import os
+
+from enum import Enum
+
+# NB: Avoid using "Optional" and use "| None" instead, as "Option" and "Optional" look similar.
+from typing import (
+    Any,
+    ClassVar,
+    Dict,
+    Generic,
+    Type,
+    TypeVar,
+    cast,
+    overload,
+    TYPE_CHECKING,
+)
+
+
+from pants.option.errors import ParseError
+
+
+if TYPE_CHECKING:
+    from pants.option.subsystem import Subsystem
+
+SubsystemT = TypeVar("SubsystemT", bound="Subsystem")
+OptionT = TypeVar("OptionT", bound="Option")
+PropType = TypeVar("PropType")
+StrT = TypeVar("StrT", str, "str | None")
+IntT = TypeVar("IntT", int, "int | None")
+FloatT = TypeVar("FloatT", float, "float | None")
+BoolT = TypeVar("BoolT", bool, "bool | None")
+EnumT = TypeVar("EnumT", bound=Enum)
+# NB: We don't provide constraints, as our `XListOption` types act like a set of contraints
+ListMemberType = TypeVar("ListMemberType")
+
+
+# ================================================================
+# These will be "duplicated" with custom_types until we replace that completely.
+
+
+def _normalize_directory_separators(s: str) -> str:
+    """Coalesce runs of consecutive instances of `os.sep` in `s`, e.g. '//' -> '/' on POSIX.
+
+    The engine will use paths or target addresses either to form globs or to string-match against, and
+    including the directory separator '/' multiple times in a row e.g. '//' produces an equivalent
+    glob as with a single '/', but produces a different actual string, which will cause the engine to
+    fail to glob file paths or target specs correctly.
+
+    TODO: give the engine more control over matching paths so we don't have to sanitize the input!
+    """
+    return os.path.normpath(s)
+
+
+class target_option(str):
```

Comment:
These don't have to be classes strictly speaking, but it is really convenient if they are.
(Them being a class sets the right metavar, handles the conversion without duplicating between the singular and list types, and allows us to just "call" the type for conversion)

Reply:
Nevermind, fixed it up to use existing functions.